### ISSUE TITLE
Retrieving (older) logs via livestatus has issues

### ISSUE THREAD

**Issue by GuusH** [naparuba/shinken#1566]
Hi,

We're trying to use the reporting functionality of Thruk on top of our Shinken installation. We've found that in both 2.0.3 or 2.2 if you request logs that go further back than a few weeks, the broker usually chokes on it. We're storing our logs in Mongo using the appropriate broker modules. In order to work around this, we've tried to implement Thruks logcache functionality, but in order to fill the cache, it first needs to request all the logs via the broker, which fails. So catch 22.

Some details. We've been working on this in our test environment, which is fairly small. In runs on a couple of VMs and monitors about 10 hosts and 350 services. The logs in Mongo are about 6 GB:

    opsrs_dev_1:PRIMARY> show dbs;
    local	1.078125GB
    logs	5.951171875GB
    scheduler	0.953125GB

The VM that runs the broker had 2 GB of RAM. When requesting the logs (either by running a report in Thruk that requests logs for last quarter or by running the import via the commandline 'thruk -a logcacheimport --local'), the main broker process starts using about 100% and quickly eats up all memory and starts to swap. After a while the box trashes itself to a standstill. I upgraded the VM to 4 GB RAM: same story. Then to 10 GB. In this case the broker eats about 9 GB of RAM and the system barely swaps. Until the broker reaches about the 9 GB RAM usage mark, it stays responsive on the livestatus module (Thruk stays working), but after that it stops responding, which results in various error messages in Thruk. The broker keeps on churning 100% CPU and 93% RAM, but nothing seems to be actually happening. I've even left it in that state overnight just to see what would happen, but it was in the same state this morning. After restarting the broker and reloading Shinken the environment started working again, without completing the logs request.

I haven't looked at the code, but I hope the broker doesn't need to read every log line before it can start returning results to the requestor. Looks like we have about 14 GB of logs in production, so I haven't dared to try requesting all logs from there. I do know that if I run the availability report in Thruk for last quarter for about 20 hosts (out of 1300 total) it throws an ugly error after several minutes, so I'm pretty sure it's not working in 2.0.3 either.

Any idea what's going on here?

**Comment by naparuba** [comment]
it's a nown limitation since years, and if I'm nto wrong as far the livestatus protocol don' change here for paging system or such things, there still be there.

The log table is here for compatibility with the C implementation of livestauts, but so it inherit from the same conception fails.

That's why I always to my customers to NOt use logs inside livestauts module. never.

You can see with thruk/multisite authors if it's possible that they do paging queries or things like this, but until the official livestatus proctool is not changed, we cannot change the module behavior.

**Comment by sni** [comment]
i wouldn't say this is a limitation of the livestatus protocol. Livestatus supported paging from day 0 and Thruk makes heavy use of that. Also Thruk does not import everything at once, but instead imports one day at a time. (This can be changed by parameters of course)
However, i think the worst problem here, is that mongodb (although its focus is big data) just can't handle lots of data and trys to put everything in memory, which is totally useless for logfile of course. This is the
#1 reason why Thruk changed to mysql logfile cache some years ago.
Btw, Thruk also supports import directly from the logfiles, see http://thruk.org/documentation/logfile-cache.html for some examples. At least for the initial import this could be a good idea.

**Comment by GuusH** [comment]
Thanks for chiming in. In this case it's definitely not Mongo that's blowing up, but the Shinken broker. I definitely need to look at the livestatus module code what's happening, because I still think something weird is going on.

**Comment by GuusH** [comment]
I finally had a little time to look at this. Working my way forward from the logstore MongoDB module. I've put some debug stuff in place and eventhough I've created a hostgroup availability report, it gets every record from MongoDB, including logentries for services. So now I'm sure that it gets the wrong records and the wrong amount of records. Possible culprits: MongoDB itself (I'm on 2.4.9) or the piece of code that generates the filter. I'm going to upgrade MongoDB to latest stable (2.6.x, test again, then 3.x, test again) to see if that makes a difference. If that doesn't help, I'm going to look at the code that generates the filter, which determines the records it gets from Mongo. To be continued...

**Comment by gst** [comment]
>  I'm going to look at the code that generates the filter

If I were you I wouldn't do that.. are at least not on a monday at the begin of the month ;)

**Comment by GuusH** [comment]
So I found the problem. Thruk requests a Filter attribute ('message') that the mod-logstore-mongodb module doesn't know how to handle and puts a dummy entry in place, which causes it to request way too many logs. The issue referenced a few minutes ago explains it in detail and links a PR to solve the problem.
While troubleshooting I upgraded all the way to MongoDB 3 and uploaded PRs for the mod-logstore-mongodb and mod-retention-mongodb modules to make them MongoDB 3 (and pymongo3) compatible. PRs are available in their respective github locations.
Hope you find this useful!
